File: Days 31 - 40/Day 38/ValidQueenSetups.cpp
#include <cmath>
#include <iostream>
#include <vector>

bool IsValidQueenSetup(const std::vector<unsigned int>& board)
{
	unsigned int lastRow = board.size() - 1;
	unsigned int lastColumn = board[board.size() - 1];

	for (unsigned int i = 0; i < lastRow; i++)
	{
		unsigned int difference = std::abs(static_cast<int>(lastColumn - board[i]));

		if ((difference == 0) || (difference == lastRow - i))
		{
			return false;
		}
	}

	return true;
}

unsigned int GetValidQueenSetupsHelper(std::vector<unsigned int>& board, const unsigned int N)
{
	if (board.size() == N)
	{
		return 1;
	}

	unsigned int count = 0;

	for (unsigned int i = 0; i < N; i++)
	{
		board.push_back(i);

		if (IsValidQueenSetup(board))
		{
			count += GetValidQueenSetupsHelper(board, N);
		}

		board.pop_back();
	}

	return count;
}

unsigned int GetValidQueenSetups(const unsigned int N)
{
	std::vector<unsigned int> board;

	return GetValidQueenSetupsHelper(board, N);
}

int main(int argc, const char* argv[])
{
	std::cout << "Total solutions for N = 2: " << GetValidQueenSetups(2) << "\n";
	std::cout << "Total solutions for N = 4: " << GetValidQueenSetups(4) << "\n";
	std::cout << "Total solutions for N = 8: " << GetValidQueenSetups(8) << "\n";

	std::cin.get();

	return 0;
}

File: Days 31 - 40/Day 38/ValidQueenSetups.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day38
	{
		private static int Main(string[] args)
		{
			Console.WriteLine($"Total solutions for N = 2: {GetValidQueenSetups(2)}");
			Console.WriteLine($"Total solutions for N = 4: {GetValidQueenSetups(4)}");
			Console.WriteLine($"Total solutions for N = 8: {GetValidQueenSetups(8)}");

			Console.ReadLine();

			return 0;
		}

		private static int GetValidQueenSetups(int n)
		{
			List<int> board = new List<int>();

			return GetValidQueenSetupsHelper(board, n);
		}

		private static int GetValidQueenSetupsHelper(List<int> board, int n)
		{
			if (board.Count == n)
			{
				return 1;
			}

			int count = 0;

			for (int i= 0; i < n; i++)
			{
				board.Add(i);
				
				if (IsValidQueenSetup(board))
				{
					count += GetValidQueenSetupsHelper(board, n);
				}

				board.RemoveAt(board.Count - 1);
			}

			return count;
		}

		private static bool IsValidQueenSetup(List<int> board)
		{
			int lastRow = board.Count - 1;
			int lastColumn = board[lastRow];

			for (int i = 0; i < lastRow; i++)
			{
				int difference = Math.Abs(lastColumn - board[i]);

				if (difference == 0 || difference == lastRow - i)
				{
					return false;
				}
			}

			return true;
		}
	}
}